RecoHGCal/GraphReco/plugins/peprCandidateFromHitProducer.cc

```diff
@@ -0,0 +1,345 @@
+    edm::EDGetTokenT<edm::View<reco::Track>> tracksToken_;
+    edm::EDGetTokenT<std::vector<SimCluster>> simClusterToken_;
+
+    //FIXME, to be replaced
+    bool batchedModel_;
+    std::string tritonPath_;
+    std::string inpipeName_;
+    std::string outpipeName_;
+
+    // rechit tools
+    hgcal::RecHitTools recHitTools_;
+
+    // windows
+    std::vector<InferenceWindow> windows_;
+
+    double minEta_;
+    double maxEta_;
+    double etaFrameWidth_;
+    double phiFrameWidth_;
+    size_t nEtaSegments_;
+    size_t nPhiSegments_;
+
+
+};
+
+
+peprCandidateFromHitProducer::peprCandidateFromHitProducer(const edm::ParameterSet& config) :
+        recHitCollections_(config.getParameter<std::vector<edm::InputTag> >("recHitCollections")), 
+        tracksToken_(consumes<edm::View<reco::Track>>(config.getParameter<edm::InputTag>("tracks"))),
+        simClusterToken_(consumes<std::vector<SimCluster>>(config.getParameter<edm::InputTag>("simClusters"))), 
+        batchedModel_(config.getParameter<bool>("batchedModel")), 
+        tritonPath_(config.getParameter<std::string>("tritonPath")), 
+        inpipeName_(config.getParameter<std::string>("inpipeName")),
+        outpipeName_(config.getParameter<std::string>("outpipeName")),
+        //FIXME: actually these are all not needed if windows are created in the constructor!
+        minEta_(config.getParameter<double>("minEta")),
+        maxEta_(config.getParameter<double>("maxEta")),
+        etaFrameWidth_(config.getParameter<double>("etaFrameWidth")),
+        phiFrameWidth_(config.getParameter<double>("phiFrameWidth")),
+        nEtaSegments_((size_t)config.getParameter<uint32_t>("nEtaSegments")),
+        nPhiSegments_((size_t)config.getParameter<uint32_t>("nPhiSegments"))
+        {
+
+    // sanity checks for sliding windows
+
+    // get tokens
+    for (edm::InputTag& recHitCollection : recHitCollections_) {
+        recHitTokens_.push_back(consumes<HGCRecHitCollection>(recHitCollection));
+    }
+   
+
+    produces<reco::PFCandidateCollection>();
+    //produces<std::vector<Trackster>>();
+
+
+    //launch scripts to work with Triton clients
+    //https://github.com/cms-pepr/HGCalML/tree/master/triton
+    
+    std::string str0 = "cd " + tritonPath_;
+    const char *cdcommand = str0.c_str();
+    system(cdcommand);
+
+    std::string str1 = "./cmssw_oc_server.sh > serverlog.txt &";
+    const char *clientcommand1 = str1.c_str();
+    std::cout << "Executing: " << clientcommand1 << std::endl;
+    system(clientcommand1); 
+
+    std::cout << "Sleeping 30s... " << std::endl;
+    system("sleep 30");
+
+    std::string str2 = "./cmssw_oc_forward_client.sh " + inpipeName_;
+    const char *clientcommand2 = str2.c_str();
+    std::cout << "Executing: " << clientcommand2 << std::endl;
+    system(clientcommand2);
+
+    system("cd -");
+
+}
+
+peprCandidateFromHitProducer::~peprCandidateFromHitProducer() {
+
+    //FIXME: find process ID of scripts run in the constructor
```

Comment:
If you start the scripts with this helper function you get their PIDs:

#include <sys/types.h>
#include <unistd.h>
#include <string>
#include <cstdlib>

pid_t start_background(std::string osstr)
{
   int   rv;
   pid_t child_pid= fork();

   if(child_pid)
   {
       return child_pid;
   }
   else
   {
       system(osstr.data());
   }

   return(child_pid);
}

Reply:
I think it is not working yet; as I mentioned the producer seems to be running twice. In any case, until it is resolved one should just be able to start the client manually (but then take care the correct pipenames are used).

Reply:
Now starting the forward client in background works.I added a check if the pipe is open before writing to the pipe; this prevents the producer from proceeding before the client is ready to accept data.

Reply:
Killing the client script in the destructor should still be fixed. The current (commented) attempt makes the child process crash in an ugly way, while the client keeps running.

Reply:
I think I got the solution, not very elegant, but should work!

std::string tempfilename=/dev/shm/<pid_or_whatever_identifier>
system(("<client script> & echo $! > "+tempfilename).data());
int childpid = <read tempfilename>

Reply:
Works now with the Triton client script put in the GraphReco/test directory